### PR TITLE
Move SeqWeight out of the sequence distributions

### PR DESCRIPTION
**Pull request by aarmey**
This is far simpler, and properly renormalizes the relative distribution weights.

### CODE REVIEW

File: msresist/binomial.py [hunk 1]
```diff
@@ -164,13 +163,13 @@ def from_summaries(self, inertia=0.0):
         betaA = np.sum(self.weightsIn) - k
         betaA = np.clip(betaA, 0.01, np.inf)
         probmat = sc.betainc(betaA, k + 1, 1 - self.background[0])
-        self.logWeights[:] = self.SeqWeight * np.log(np.tensordot(self.background[1], probmat, axes=2))
-        self.logWeights[:] = self.logWeights - np.mean(self.logWeights)
+        self.logWeights[:] = np.log(np.tensordot(self.background[1], probmat, axes=2))
+        self.logWeights[:] = self.logWeights - np.amax(self.logWeights)
```
Comment (aarmey) [inline msresist/binomial.py:167]: `amax` keeps the probabilities from being really small. This seems to help. I think there might be a conversion point in pom where a small LL is turned into 0.

File: msresist/expectation_maximization.py [hunk 2]
```diff
@@ -37,6 +37,11 @@ def EM_clustering(data, info, ncl, seqDist=None, gmmIn=None):
     # Add a dummy variable for the sequence information
     d = np.hstack((d, idxx.T))
 
+    # Setup weights for distributions
+    seqWarr = np.ones(d.shape[1])
+    seqWarr[-1] = seqWeight
+    seqWarr /= np.sum(seqWarr)
```
Comment (aarmey) [inline msresist/expectation_maximization.py:43]: This means that the sequence contribution doesn't explode. When the sequence contribution becomes large (e.g., `seqWeight = 100.0`, the Gaussians are scaled down appropriately.

File: msresist/expectation_maximization.py [hunk 3]
```diff
@@ -58,13 +63,13 @@ def EM_clustering(data, info, ncl, seqDist=None, gmmIn=None):
                 nDist[-1].summarize(d[:, -1], weights=weights)
                 nDist[-1].from_summaries()
 
-                dists.append(IndependentComponentsDistribution(nDist))
+                dists.append(IndependentComponentsDistribution(nDist, weights=seqWarr))
```
Comment (aarmey) [inline msresist/expectation_maximization.py:66]: ICD can handle seqWeights for us.

File: msresist/expectation_maximization.py [hunk 4]
```diff
@@ -58,13 +63,13 @@ def EM_clustering(data, info, ncl, seqDist=None, gmmIn=None):
...
 
             gmm = GeneralMixtureModel(dists)
         else:
             gmm = gmmIn
 
-        gmm.fit(d, max_iterations=200, verbose=False, stop_threshold=1e-6)
+        gmm.fit(d, max_iterations=2000, verbose=False, stop_threshold=1e-9)
```
Comment (aarmey) [inline msresist/expectation_maximization.py:72]: I think because of a reduced roundoff error, fitting seems to be able to continue for longer.

**Comment by mcreixell** [comment]
I've been playing with this using the AXL data set and it seems to assign all peptides to only one cluster every time. 

Screen Shot 2021-03-11 at 10 42 36 AM

File: msresist/expectation_maximization.py (hunk 2 above)
Comment (mcreixell) [reply msresist/expectation_maximization.py:43]: this provides a weight per feature (e.g. 10 AXL mutant cell lines), right? if so why does the last cell line get `seqWeight`?

Comment (aarmey) [reply msresist/expectation_maximization.py:43]: Because the sequence distribution is tacked onto the end as an additional "feature".

**Comment by aarmey** [comment]
@mcreixell yes, the fixed case is behaving weird. Are the PSPLs normalized so that a random peptide would on average be just as close to each of them?

File: msresist/expectation_maximization.py (hunk 2 above)
Comment (mcreixell) [reply msresist/expectation_maximization.py:43]: Oh true got it.

**Comment by mcreixell** [comment]
The PSPLs are log2 transformed but there are some differences between profilings generated by different labs... I tried to normalize them equally but it's not perfect.

**Comment by aarmey** [comment]
Looking at the fixed case a bit more.

**Comment by aarmey** [comment]
Try it now.

**Comment by mcreixell** [comment]
Now it works better but one of the clusters is always empty

**Comment by aarmey** [comment]
Sure. I’ll think about this more, too. In the mean time, try the other methods.

**Comment by mcreixell** [comment]
The regular PAM250 method works great but binomial assigns all peptides to the same cluster.

**Comment by aarmey** [comment]
That's strange... the only change is how seqWeight is handled, so I don't see how there could be a PAM250/binomial difference. Also, everything is identical to before with seqWeight=0...

**Comment by aarmey** [comment]
Are clusters truly empty, or only empty after hard assignment?

**Comment by mcreixell** [comment]
After hard assignment.

**Comment by mcreixell** [comment]
With binomial if SeqWeight is set to 10,000 there are no empty clusters whereas if it's a small weight (<100) only one ore two clusters have peptides.

**Comment by mcreixell** [comment]
Do you think we'll make significant changes to the model? I'd like to start re-pickling fittings.

**Comment by aarmey** [comment]
I don't anticipate doing so, so long as this is working.

**Comment by mcreixell** [comment]
The only problem I see is getting empty clusters after hard assignment when `SeqWeight` is 0 but that's with the AXL data set and maybe it was the case before and I don't remember. I'm going to try wit the MCF7 data set that is larger but not as large as CPTAC.

**Comment by aarmey** [comment]
Ok, let me know if it keeps happening, but I really think the behavior must be the same as before.

**Comment by mcreixell** [comment]
I'm replacing all the pickled models. As soon as I have these I will commit and merge.

**Review by mcreixell — APPROVED** [review]

**Comment by mcreixell** [comment]
I realized PCA taking forever to run with CPTAC...  I guess because of the overall size of the data set and the amount of missing data. Aretha ends up breaking (VSCode asks me to reload) and haven't been able to fit the model to CPTAC once. Is there an argument within PCA we could play with? Maybe tolerance? @aarmey 

https://github.com/meyer-lab/resistance-MS/blob/39b22e7e835fcf2cb1c6cdcd0333a287d8da4f8c/msresist/expectation_maximization.py#L31

**Comment by aarmey** [comment]
Weird... how can I run the fitting?

**Comment by mcreixell** [comment]
I made a figure files for each pam250 model with a different weight: `make figurePAM20.svg`

**Comment by aarmey** [comment]
@mcreixell I switched the PCA method from SVD to NIPALS which seems to have fixed it. First time I've had something large enough to make a difference! (I guess the other factor is this is running PCA 100 times to do the imputation.)

**Comment by mcreixell** [comment]
Oh great! I didn't know you could change the method of sklearn's PCA, it doesn't appear on the documentation. Thanks.

**Comment by aarmey** [comment]
https://www.statsmodels.org/stable/generated/statsmodels.multivariate.pca.PCA.html

**Comment by mcreixell** [comment]
Oh that's because it's not sklearns PCA... nvm

**Comment by aarmey** [comment]
By the way, once it gets past PCA it uses much, much less cpu, so you can probably run 4–5 fittings in parallel.

**Comment by mcreixell** [comment]
Ok great that's helpful.

**Comment by mcreixell** [comment]
It looks like we're only getting nan values as soon as we use ~15 or more. It works with around 10 clusters.

**Comment by aarmey** [comment]
I had an error—if you initialize the Gaussian with hard assignments, and all the assigned peptides are missing, you start with NaN. I softened the initial assignments a bit.

**Comment by aarmey** [comment]
I can't get this to fail in the tests even when I removed the repeat-on-NaN check, so hopefully this behaves.

**Comment by mcreixell** [comment]
Now it works wonderfully with 30 clusters, currently trying with a few more.

**Comment by mcreixell** [comment]
It seems to work even with 50 clusters now

**Comment by aarmey** [comment]
@mcreixell as this is building, shall I review and merge?

**Comment by aarmey** [comment]
Oh wait you have it on a reduced set of figures... nvm.

**Comment by mcreixell** [comment]
I'm going to include a couple model files in this PR (1 ready, 1 still running) but then I'd like to merge this and open a new PR to work on the imputation figure and then update all figures with whatever models works best according to imputation and predictive performance.

**Comment by mcreixell** [comment]
I unpickled a model with 40 clusters and strangely the centers are the same for each cluster.

Screen Shot 2021-03-15 at 3 12 49 PM

I checked with the AXL data set and this is not the case. 

Screen Shot 2021-03-15 at 3 15 07 PM

**Comment by mcreixell** [comment]
Oh I just realized the values are not exactly the same

**Comment by mcreixell** [comment]
Maybe I'm using too many clusters? I'll try with different number of clusters and weights in a new PR.

**Comment by aarmey** [comment]
Reducing the minimum stdev might also help, since that will allow distributions to "focus" on specific values.

**Comment by mcreixell** [comment]
I reduced the minimum stdev to 0.01 and tried with 40, 35, 30, 25, and 20 clusters and all of them give identical centers across clusters... Any ideas of what might be happening? This doesn't happen at all with the AXL data set. It either fits great or returns breaks the nan assertion but never returns this identical centers. @aarmey

**Comment by aarmey** [comment]
Does this happen with SeqWeight=0?

**Comment by mcreixell** [comment]
Yes I tried different weights as well and the same happens.

**Comment by mcreixell** [comment]
I tried with binomial as well and in this case the second iteration is always worse:

Screen Shot 2021-03-17 at 8 22 43 AM

**Comment by aarmey** [comment]
Even with SeqWeights=0? More information is always helpful here.

**Comment by mcreixell** [comment]
No, when weight is 0 happens the same as with PAM250. It fits but centers are the same for each cluster.

**Comment by aarmey** [comment]
Can you check if the clusters are the same if you skip the fit step (i.e. return the model with just the k-means init)? Also check if you run the fitting for only one iteration.

**Comment by mcreixell** [comment]
With the binomial method or with both?

**Comment by mcreixell** [comment]
First cell shows kmeans centers and the second one DDMC centers after 1 iteration:

Screen Shot 2021-03-17 at 3 16 30 PM

**Comment by mcreixell** [comment]
Even with 5 clusters the same happens.

**Comment by aarmey** [comment]
It could be helpful to quantify this in some way, to make sure it's not just the data on the end that you're printing.

Are the scores also flat?

I'm not sure why it's happening, but I can't see how this would arise from the changes in this PR. Does it happen with fitting on master? Maybe there were some changes made before and we just didn't notice because we've been pickling the model.

You could also try manually stepping through the fitting process. `fit` is basically working back and forth between `gmm.predict_proba()` and `dists[ii].fit(d, weights=weights)`. Watching each step, maybe it's clear that one or the other suddenly flattens everything out.

**Comment by mcreixell** [comment]
The scores look flat across samples (2nd cell) while centers are flat across clusters (1st cell):
Screen Shot 2021-03-17 at 4 52 05 PM

The clusters we're using in master look good (below) but I agree I don't see what in this PR could be causing this to happen...
Screen Shot 2021-03-17 at 4 44 45 PM

I'm currently working on watching each step.

**Comment by mcreixell** [comment]
@aarmey just to make sure I understand correctly, line 62 here fits only the sequence distribution using the k-means labels because the rest of the data was already initialized by k-means, right?

https://github.com/meyer-lab/resistance-MS/blob/307f37e6b0e9e10d9bbb896893374747bf0a5bc7/msresist/expectation_maximization.py#L51-L62

**Comment by mcreixell** [comment]
The weirdest thing is that this also happens even with the entire portion of the data set that only contains 1.3k peptides instead of 30k.

**Comment by aarmey** [comment]
`IndependentComponentsDistribution` contains all of the distributions, so the one `fit()` call per cluster is updating everything to match k-means.

Do you know if `GMM.fit()` starts with M or E? Because if it does M first, it might be overwriting this initial setup based on whatever initial scores it uses...

**Comment by aarmey** [comment]
The other reason that it has to be something related to `GMM.fit()` is they aren't all the same when you output right after k-means initialization...

**Comment by aarmey** [comment]
It looks like this might be it, but that we can provide the initial clusters as the `weights` argument to `GMM.fit()`. I can hopefully play with it a little bit tomorrow morning...

**Comment by aarmey** [comment]
@mcreixell I just made a quick hack that I think accomplishes this. Try it out.

**Comment by aarmey** [comment]
This seems to have also fixed Binomial exiting after one iteration.

**Comment by mcreixell** [comment]
Binomial now doesn't exit after the first iteration but I'm still seeing the same centers per sample with both methods...

**Comment by mcreixell** [comment]
Don't these gmm centers before fitting look weird tho? The values progressively increase per row.

kmeans centers and gmm centers before fitting:
Screen Shot 2021-03-18 at 9 32 15 AM

gmm centers after gmm.summarize() and after fitting:
Screen Shot 2021-03-18 at 9 32 45 AM

**Comment by aarmey** [comment]
Up above, when you checked master, you unpickled a model. Don't do that—look at fitting anew.

**Comment by aarmey** [comment]
Interesting... the initialization did change in this PR, so let me set it up another way...

**Comment by mcreixell** [comment]
These prints are coming from a notebook where I'm fitting a new model with just the complete portion of the CPTAC data

**Comment by aarmey** [comment]
So you think the k-means results themselves look weird?

Let's switch to Slack.

**Comment by mcreixell** [comment]
Now one of the clusters is different

Screen Shot 2021-03-18 at 10 38 06 AM

**Comment by aarmey** [comment]
Had a realization about this. There are four changes here:
1. Moving seqWeight out of the distance classes.
2. Initializing differently.
3. Scaling the probabilities by the maximum instead of the average.
4. Setting `std_min` in the Gaussian distributions.

Only the first makes it necessary to repickle the fits, and I think the others are the things that matter. Why don't you make separate PR's for the third and fourth changes, then see how those behave? They are each only one or two line changes, and I think might be what is avoiding nan results. If it's not actually helping, we don't need to do the first. Finally, this process will help isolate which of these is causing the clusters to be similar.

**Comment by aarmey** [comment]
@mcreixell do you need to keep anything on this branch or can I delete it?

**Comment by mcreixell** [comment]
I don't think so.

**Pull request closed without merging**